File: sdks/go/hearth/middleware.go
package hearth

import (
	"net/http"
)

// MiddlewareConfig configures a RequirePermission middleware factory.
type MiddlewareConfig struct {
	// ExpectedMode is required. Controls how the middleware evaluates permissions.
	//   ModeEmbedded      — decodes JWT locally; no network call.
	//   ModeIntrospection — calls POST /introspect; requires ClientID + ClientSecret.
	//   ModeDecision      — calls POST /oauth/authorize; fail-closed on network errors.
	//
	// The middleware never silently falls back from one mode to another regardless
	// of what claims are present in the token.
	ExpectedMode AccessTokenAuthorizationMode

	// ClientID and ClientSecret are required when ExpectedMode is ModeIntrospection.
	ClientID     string
	ClientSecret string

	// TokenExtractor extracts the bearer token from an incoming request.
	// Defaults to extracting from the Authorization: Bearer <token> header.
	TokenExtractor func(r *http.Request) string

	// OnDenied is called when the middleware rejects a request due to missing
	// permission (403 Forbidden). Defaults to writing HTTP 403.
	OnDenied func(w http.ResponseWriter, r *http.Request)

	// OnUnauthorized is called when the middleware rejects a request due to an
	// invalid token — session-version revoked, cache stale, or inactive token
	// (401 Unauthorized). Defaults to writing HTTP 401.
	//
	// Note: when cfg.ExpectedMode is ModeEmbedded and WithSessionVersions was
	// configured on the client, sv-related rejections call OnUnauthorized, not
	// OnDenied, to match the 401 semantics prescribed by RFC HEA-930 § 8.
	OnUnauthorized func(w http.ResponseWriter, r *http.Request)
}

// bearerToken is the default TokenExtractor: strips "Bearer " prefix.
func bearerToken(r *http.Request) string {
	auth := r.Header.Get("Authorization")
	if len(auth) > 7 && auth[:7] == "Bearer " {
		return auth[7:]
	}
	return ""
}

// forbiddenHandler is the default OnDenied handler.
func forbiddenHandler(w http.ResponseWriter, _ *http.Request) {
	http.Error(w, "Forbidden", http.StatusForbidden)
}

// unauthorizedHandler is the default OnUnauthorized handler.
func unauthorizedHandler(w http.ResponseWriter, _ *http.Request) {
	http.Error(w, "Unauthorized", http.StatusUnauthorized)
}

// RequirePermission returns a middleware that enforces a permission check before
// calling next. The check strategy is controlled by cfg.ExpectedMode.
//
// Behavior per mode:
//   - ModeEmbedded:      decodes JWT claims locally; no network round-trip.
//     When the client has a SessionVersionCache configured, the sv claim is
//     validated first (RFC HEA-930 § 8). Revoked or stale → OnUnauthorized (401).
//   - ModeIntrospection: calls POST /introspect, verifies the echoed mode matches
//     cfg.ExpectedMode, then checks the live permissions claim. Returns
//     ModeMismatchError (mapped to denial) when modes disagree.
//   - ModeDecision:      calls POST /oauth/authorize; fail-closed on any error.
//
// The middleware NEVER silently falls back from decision→embedded or
// introspection→embedded based on whether `permissions` is absent in the token.
func RequirePermission(c *Client, permission string, cfg MiddlewareConfig) func(http.Handler) http.Handler {
	extract := cfg.TokenExtractor
	if extract == nil {
		extract = bearerToken
	}
	deny := cfg.OnDenied
	if deny == nil {
		deny = forbiddenHandler
	}
	unauth := cfg.OnUnauthorized
	if unauth == nil {
		unauth = unauthorizedHandler
	}

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			token := extract(r)
			if token == "" {
				deny(w, r)
				return
			}

			var allowed bool

			switch cfg.ExpectedMode {
			case ModeEmbedded:
				claims := decodeClaims(token)
				if claims == nil {
					deny(w, r)
					return
				}
				// Session-version check when cache is configured (RFC HEA-930 § 8).
				if c.svCache != nil {
					svPresent := claims.SV != nil
					var sv uint64
					if svPresent {
						sv = *claims.SV
					}
					result := c.svCache.Check(svPresent, sv, claims.Sid)
					switch result {
					case SvRevoked:
						unauth(w, r)
						return
					case SvStale:
						unauth(w, r)
						return
					case SvOK, SvSkip:
						// continue to permission check
					}
				}
				allowed = contains(claims.Permissions, permission)

			case ModeIntrospection:
				resp, err := c.Introspect(r.Context(), IntrospectRequest{
					Token:        token,
					ClientID:     cfg.ClientID,
					ClientSecret: cfg.ClientSecret,
				})
				if err != nil {
					deny(w, r)
					return
				}
				if !resp.Active {
					deny(w, r)
					return
				}
				// Verify the echoed mode matches the configured expectation.
				// An absent mode field defaults to "embedded" (server omits it for Embedded clients).
				echoedMode := resp.Mode
				if echoedMode == "" {
					echoedMode = string(ModeEmbedded)
				}
				if echoedMode != string(cfg.ExpectedMode) {
					// Hard rejection — do not fall back to local check.
					deny(w, r)
					return
				}
				allowed = contains(resp.Permissions, permission)

			case ModeDecision:
				// CheckPermission is fail-closed: network errors return Allowed=false.
				resp, _ := c.CheckPermission(r.Context(), token, CheckPermissionRequest{
					Permission: permission,
				})
				allowed = resp != nil && resp.Allowed

			default:
				// Unknown mode is a misconfiguration — deny.
				deny(w, r)
				return
			}

			if !allowed {
				deny(w, r)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

File: sdks/python/src/hearth/middleware.py
"""Mode-aware ASGI and WSGI permission middleware for Hearth.

Both ``RequirePermissionMiddleware`` (ASGI) and ``WsgiPermissionMiddleware``
(WSGI) enforce an explicit ``mode`` parameter — they MUST NOT auto-detect the
mode from JWT claim presence.  Per spec §15.3: absence of a ``permissions``
claim in the token does NOT trigger a fallback to a network mode.

Usage (ASGI / Starlette / FastAPI)::

    from hearth import HearthClient
    from hearth.middleware import RequirePermissionMiddleware

    client = HearthClient(base_url, realm_id=realm_id)
    app = RequirePermissionMiddleware(
        app,
        client=client,
        permission="docs.write",
        mode="embedded",          # "embedded" | "introspection" | "decision"
    )

Usage (WSGI / Flask / Django)::

    from hearth.middleware import WsgiPermissionMiddleware

    app.wsgi_app = WsgiPermissionMiddleware(
        app.wsgi_app,
        client=client,
        permission="docs.write",
        mode="decision",
    )
"""

from __future__ import annotations

import asyncio
from typing import Awaitable, Callable, Optional, TYPE_CHECKING

from .claims import Claims
from .errors import AuthorizationModeMismatchError, RequiredActionError

if TYPE_CHECKING:
    from .client import HearthClient

# ASGI/WSGI callable type aliases (not enforced at runtime, only for type checkers).
ASGIApp = Callable[..., Awaitable[None]]
WSGIApp = Callable[..., object]


# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------

def _extract_bearer_asgi(scope: dict) -> Optional[str]:
    """Return the Bearer token from an ASGI HTTP scope, or None."""
    for name, value in scope.get("headers", []):
        if name.lower() == b"authorization":
            auth = value.decode("latin-1")
            if auth.startswith("Bearer "):
                return auth[7:]
    return None


def _extract_bearer_environ(environ: dict) -> Optional[str]:
    """Return the Bearer token from a WSGI environ dict, or None."""
    auth = environ.get("HTTP_AUTHORIZATION", "")
    if auth.startswith("Bearer "):
        return auth[7:]
    return None


def _check_embedded(token: str, permission: str) -> bool:
    """Decode JWT locally and check ``permissions`` claim.

    Returns ``False`` when the claim is absent — never falls back to a network
    mode (design constraint: absence of claim ≠ switch mode).
    """
    try:
        claims = Claims.decode(token)
        perms = claims.get("permissions") or []
        return permission in perms
    except Exception:
        return False


def _check_introspection_sync(
    client: "HearthClient",
    token: str,
    permission: str,
    client_id: str,
    client_secret: Optional[str],
    expected_mode: str,
) -> bool:
    """Call POST /introspect, validate the echoed mode, then check the permission.

    :raises AuthorizationModeMismatchError: when the server echoes a mode that
        differs from ``expected_mode``.  Callers should map this to a denial.
    """
    resp = client.introspect(token, client_id=client_id, client_secret=client_secret)
    if not resp.active:
        return False
    # An absent mode field defaults to "embedded" (server omits it for embedded clients).
    echoed = resp.mode or "embedded"
    if echoed != expected_mode:
        raise AuthorizationModeMismatchError(expected_mode, echoed)
    return permission in (resp.permissions or [])


def _is_required_action_token(token: str) -> bool:
    """Return True when the token's token_type claim is 'required_action'."""
    try:
        return Claims.decode(token).token_type() == "required_action"
    except Exception:
        return False


async def _send_401_required_action(send: Callable) -> None:
    """Emit a minimal HTTP 401 ASGI response for required-action tokens (spec §6 rule 6)."""
    await send({
        "type": "http.response.start",
        "status": 401,
        "headers": [
            [b"content-type", b"text/plain; charset=utf-8"],
            [b"www-authenticate", b'Bearer realm="hearth", error="required_action"'],
        ],
    })
    await send({"type": "http.response.body", "body": b"Required actions pending"})


async def _send_403(send: Callable) -> None:
    """Emit a minimal HTTP 403 ASGI response."""
    await send({
        "type": "http.response.start",
        "status": 403,
        "headers": [[b"content-type", b"text/plain; charset=utf-8"]],
    })
    await send({"type": "http.response.body", "body": b"Forbidden"})


def _wsgi_401_required_action(start_response: Callable) -> list:
    """Return a minimal HTTP 401 WSGI response for required-action tokens (spec §6 rule 6)."""
    start_response(
        "401 Unauthorized",
        [
            ("Content-Type", "text/plain"),
            ("WWW-Authenticate", 'Bearer realm="hearth", error="required_action"'),
        ],
    )
    return [b"Required actions pending"]


def _wsgi_403(start_response: Callable) -> list:
    """Return a minimal HTTP 403 WSGI response."""
    start_response("403 Forbidden", [("Content-Type", "text/plain")])
    return [b"Forbidden"]


# ---------------------------------------------------------------------------
# ASGI middleware
# ---------------------------------------------------------------------------

class RequirePermissionMiddleware:
    """ASGI middleware that enforces a Hearth permission check on every HTTP request.

    Non-HTTP connections (WebSocket, lifespan) pass through without checks.

    :param app: The downstream ASGI application.
    :param client: An authenticated :class:`~hearth.client.HearthClient`.
    :param permission: Permission string to require, e.g. ``"docs.write"``.
    :param mode: Authorization mode.  Must be one of ``"embedded"``,
        ``"introspection"``, or ``"decision"``.  NEVER auto-detected.
    :param client_id: Required when ``mode="introspection"``.
    :param client_secret: Optional client secret for introspection.
    :param organization_id: Optional org-scope for decision/introspection checks.
    :param resource: Optional RFC 8707 resource indicator.
    """

    def __init__(
        self,
        app: ASGIApp,
        *,
        client: "HearthClient",
        permission: str,
        mode: str,
        client_id: str = "",
        client_secret: str = "",
        organization_id: Optional[str] = None,
        resource: Optional[str] = None,
    ) -> None:
        self._app = app
        self._client = client
        self._permission = permission
        self._mode = mode
        self._client_id = client_id
        self._client_secret = client_secret or None
        self._organization_id = organization_id
        self._resource = resource

    async def __call__(self, scope: dict, receive: Callable, send: Callable) -> None:
        if scope.get("type") != "http":
            await self._app(scope, receive, send)
            return

        token = _extract_bearer_asgi(scope)
        if not token:
            await _send_403(send)
            return

        if _is_required_action_token(token):
            # Spec §6 rule 6: required_action tokens must never be accepted for
            # general API access — respond 401 and do not call next.
            await _send_401_required_action(send)
            return

        try:
            allowed = await self._check(token)
        except Exception:
            # Fail-closed on any error (mode mismatch, network, etc.)
            allowed = False

        if not allowed:
            await _send_403(send)
            return

        await self._app(scope, receive, send)

    async def _check(self, token: str) -> bool:
        if self._mode == "embedded":
            return _check_embedded(token, self._permission)

        if self._mode == "decision":
            # Run sync network call in thread pool to avoid blocking event loop.
            loop = asyncio.get_event_loop()
            result = await loop.run_in_executor(
                None,
                lambda: self._client.check_permission(
                    token,
                    self._permission,
                    organization_id=self._organization_id,
                    resource=self._resource,
                ),
            )
            return result.allowed

        if self._mode == "introspection":
            loop = asyncio.get_event_loop()
            return await loop.run_in_executor(
                None,
                lambda: _check_introspection_sync(
                    self._client,
                    token,
                    self._permission,
                    self._client_id,
                    self._client_secret,
                    self._mode,
                ),
            )

        # Unknown mode — misconfiguration. Deny.
        return False


# ---------------------------------------------------------------------------
# WSGI middleware
# ---------------------------------------------------------------------------

class WsgiPermissionMiddleware:
    """WSGI middleware that enforces a Hearth permission check on every request.

    :param app: The downstream WSGI application.
    :param client: An authenticated :class:`~hearth.client.HearthClient`.
    :param permission: Permission string to require, e.g. ``"docs.write"``.
    :param mode: Authorization mode.  Must be one of ``"embedded"``,
        ``"introspection"``, or ``"decision"``.  NEVER auto-detected.
    :param client_id: Required when ``mode="introspection"``.
    :param client_secret: Optional client secret for introspection.
    :param organization_id: Optional org-scope for decision/introspection checks.
    :param resource: Optional RFC 8707 resource indicator.
    """

    def __init__(
        self,
        app: WSGIApp,
        *,
        client: "HearthClient",
        permission: str,
        mode: str,
        client_id: str = "",
        client_secret: str = "",
        organization_id: Optional[str] = None,
        resource: Optional[str] = None,
    ) -> None:
        self._app = app
        self._client = client
        self._permission = permission
        self._mode = mode
        self._client_id = client_id
        self._client_secret = client_secret or None
        self._organization_id = organization_id
        self._resource = resource

    def __call__(self, environ: dict, start_response: Callable) -> object:
        token = _extract_bearer_environ(environ)
        if not token:
            return _wsgi_403(start_response)

        if _is_required_action_token(token):
            # Spec §6 rule 6: required_action tokens must never be accepted for
            # general API access — respond 401 and do not call next.
            return _wsgi_401_required_action(start_response)

        try:
            allowed = self._check(token)
        except Exception:
            # Fail-closed on any error (mode mismatch, network, etc.)
            allowed = False

        if not allowed:
            return _wsgi_403(start_response)

        return self._app(environ, start_response)

    def _check(self, token: str) -> bool:
        if self._mode == "embedded":
            return _check_embedded(token, self._permission)

        if self._mode == "decision":
            result = self._client.check_permission(
                token,
                self._permission,
                organization_id=self._organization_id,
                resource=self._resource,
            )
            return result.allowed

        if self._mode == "introspection":
            return _check_introspection_sync(
                self._client,
                token,
                self._permission,
                self._client_id,
                self._client_secret,
                self._mode,
            )

        # Unknown mode — misconfiguration. Deny.
        return False
